File: compatibility-tests/sdk-test-go/tests/iam_test.go
package tests

import (
	"context"
	"testing"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/iam"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestIAM(t *testing.T) {
	ctx := context.Background()
	svc := testutil.IAMClient()
	roleName := "go-test-role"
	assumePolicy := `{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}`

	t.Cleanup(func() {
		svc.DeleteRole(ctx, &iam.DeleteRoleInput{RoleName: aws.String(roleName)})
	})

	t.Run("CreateRole", func(t *testing.T) {
		_, err := svc.CreateRole(ctx, &iam.CreateRoleInput{
			RoleName:                 aws.String(roleName),
			AssumeRolePolicyDocument: aws.String(assumePolicy),
		})
		require.NoError(t, err)
	})

	t.Run("GetRole", func(t *testing.T) {
		r, err := svc.GetRole(ctx, &iam.GetRoleInput{RoleName: aws.String(roleName)})
		require.NoError(t, err)
		assert.Equal(t, roleName, aws.ToString(r.Role.RoleName))
	})

	t.Run("ListRoles", func(t *testing.T) {
		r, err := svc.ListRoles(ctx, &iam.ListRolesInput{})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Roles)
	})

	t.Run("DeleteRole", func(t *testing.T) {
		_, err := svc.DeleteRole(ctx, &iam.DeleteRoleInput{RoleName: aws.String(roleName)})
		require.NoError(t, err)
	})
}

File: compatibility-tests/sdk-test-rust/tests/iam_test.rs
mod common;

#[tokio::test(flavor = "multi_thread")]
async fn test_iam_create_role() {
    let iam = common::iam_client().await;
    let role_name = "rust-test-create-role";
    let assume_policy = r#"{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}"#;

    let _guard = common::CleanupGuard::new({
        let iam = iam.clone();
        async move {
            let _ = iam.delete_role().role_name(role_name).send().await;
        }
    });

    let result = iam
        .create_role()
        .role_name(role_name)
        .assume_role_policy_document(assume_policy)
        .send()
        .await;
    assert!(result.is_ok(), "CreateRole failed: {:?}", result.err());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_iam_get_role() {
    let iam = common::iam_client().await;
    let role_name = "rust-test-get-role";
    let assume_policy = r#"{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}"#;

    let _guard = common::CleanupGuard::new({
        let iam = iam.clone();
        async move {
            let _ = iam.delete_role().role_name(role_name).send().await;
        }
    });

    // Setup
    iam.create_role()
        .role_name(role_name)
        .assume_role_policy_document(assume_policy)
        .send()
        .await
        .expect("setup");

    let result = iam.get_role().role_name(role_name).send().await;
    assert!(result.is_ok(), "GetRole failed: {:?}", result.err());
    assert_eq!(
        result.unwrap().role().map(|r| r.role_name()).unwrap_or(""),
        role_name
    );
}

#[tokio::test(flavor = "multi_thread")]
async fn test_iam_list_roles() {
    let iam = common::iam_client().await;
    let role_name = "rust-test-list-roles";
    let assume_policy = r#"{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}"#;

    let _guard = common::CleanupGuard::new({
        let iam = iam.clone();
        async move {
            let _ = iam.delete_role().role_name(role_name).send().await;
        }
    });

    // Setup
    iam.create_role()
        .role_name(role_name)
        .assume_role_policy_document(assume_policy)
        .send()
        .await
        .expect("setup");

    let result = iam.list_roles().send().await;
    assert!(result.is_ok(), "ListRoles failed: {:?}", result.err());
    assert!(!result.unwrap().roles().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_iam_delete_role() {
    let iam = common::iam_client().await;
    let role_name = "rust-test-delete-role";
    let assume_policy = r#"{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}"#;

    // No cleanup guard needed - test is about deletion

    // Setup
    iam.create_role()
        .role_name(role_name)
        .assume_role_policy_document(assume_policy)
        .send()
        .await
        .expect("setup");

    let result = iam.delete_role().role_name(role_name).send().await;
    assert!(result.is_ok(), "DeleteRole failed: {:?}", result.err());
}
